include/mata/sparse-set.hh

```diff
@@ -0,0 +1,355 @@
+                : sparse{ std::move(other.sparse) }, dense{ std::move(other.dense) },
+                  size_{ other.size_}, capacity_{ other.capacity_} {
+            other.size_ = 0;
+            other.capacity_ = 0;
+
+            assert(consistent());
+        }
+
+        SparseSet<Number>& operator=(const SparseSet <Number>& rhs) = default;
+        SparseSet<Number>& operator=(SparseSet<Number>&& other) noexcept {
+            if (this != &other) {
+                dense = std::move(other.dense);
+                sparse = std::move(other.sparse);
+                size_ = other.size_;
+                capacity_ = other.capacity_;
+                other.size_ = 0;
+                other.capacity_ = 0;
+            }
+
+            assert(consistent());
+
+            return *this;
+        }
+
+// Things
+
+        bool consistent() {
+            return ( capacity_ >= size_ &&
+                     (max() < capacity_ || (size_ == 0 && capacity_ == 0) ) &&
+                     dense.size() >= capacity_ &&
+                     sparse.size() >= capacity_
+            );
+        }
+
+        /**
+         * @return True if predicate for @p q is set. False otherwise (even if q is out of range of the predicate).
+         */
+        bool operator[](Number q) const {
+            return has(q);
+        }
+
+        template <class InputIterator>
+        void insert(InputIterator first, InputIterator last) {
+            while (first != last) {
+                insert(*first);
+                first++;
+            }
+            assert(consistent());
+        }
+
+        template<typename T>
+        void insert_all(const T & set) {
+            insert(set.begin(),set.end());
+        }
+
+        void insert_all(const std::initializer_list<Number> & list) {
+            insert(list.begin(),list.end());
+        }
+
+        template <class InputIterator>
+        void erase(InputIterator first, InputIterator last) {
+            while (first != last) {
+                erase(*first);
+                first++;
+            }
+            assert(consistent());
+        }
+
+        template<class T>
+        void erase_all(const T & set) {
+            erase(set.begin(),set.end());
+        }
+
+        void erase_all(const std::initializer_list<Number> & list) {
+            erase(list.begin(),list.end());
+        }
+
+        //call this (instead of the fried are_disjoint) if you want the other container to be iterated (e.g. if it does not have constant membership)
+        template<class T>
+        bool intersects_with(const T & set) const {
+            for (auto i=set.begin();i<set.end();i++) {
+                if (has(*i))
+                    return true;
+            }
+            return false;
+        }
+
+        /*
+         * Complements the set with respect to a given number of elements = the maximum number + 1.
+         */
+        //probably can be done little faster
```

Comment:
So we can search for this comment later.
```suggestion
        // TODO: Probably can be done little faster.
```

Reply:
I will actually just remove this, it is not something we need to focus on.